### PR TITLE
feat: Autoregister a target driver

### PR DESCRIPTION
**Pull request by sergiyvamz**
### Summary

Automatically register the target driver in the class path

### Additional Reviewers

@karenc-bq 
@brunos-bq 
@crystall-bitquill 
@aaron-congo 
@aaronchung-bitquill 

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/using-the-jdbc-driver/UsingTheJdbcDriver.md [hunk 1]
```diff
@@ -47,19 +47,19 @@ The AWS JDBC Driver also has a parameter, [`wrapperLoggerLevel`](#aws-advanced-j
...
+| `socketTimeout`                   | `Integer` | No       | Socket timeout in milliseconds.                                                                                                                                                                                                                                                               | `null`        |
+| `tcpKeepAlive`                    | `Boolean` | No       | Enable or disable TCP keep-alive probe.                                                                                                                                                                                                                                                       | `false`       |
+| `wrapperTargetDriverAutoRegister` | `Boolean` | No       | Allows the AWS JDBC Driver to register a target driver based on `wrapperTargetDriverDialect` configuration parameter or, if it's missed, on a connection url protocol.                                                                                                                        | `true`        |
```
Comment (brunos-bq) [inline docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:62]: i don't think we should use or add the `wrapper` prefix to any property

`targetDriverAutoRegister` or even just `autoRegister`

File: wrapper/src/main/java/software/amazon/jdbc/Driver.java [hunk 2]
```diff
@@ -232,6 +214,47 @@ public DriverPropertyInfo[] getPropertyInfo(final String url, final Properties i
     return props;
   }
 
+  private java.sql.Driver getTargetDriver(
+      final @NonNull String driverUrl,
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/Driver.java:218]: NIT: `driverURL`

File: wrapper/src/main/java/software/amazon/jdbc/Driver.java [hunk 3]
```diff
@@ -232,6 +214,47 @@ public DriverPropertyInfo[] getPropertyInfo(final String url, final Properties i
...
+      final @NonNull Properties props)
+      throws SQLException {
+
+    final ConnectionUrlParser connectionUrlParser = new ConnectionUrlParser();
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/Driver.java:222]: what about
```
final ConnectionUrlParser parser = new ConnectionUrlParser();
```

File: wrapper/src/main/java/software/amazon/jdbc/ds/AwsWrapperDataSource.java [hunk 4]
```diff
@@ -259,6 +260,49 @@ ConnectionWrapper createConnectionWrapper(
         telemetryFactory);
   }
 
+  private java.sql.Driver getTargetDriver(
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/ds/AwsWrapperDataSource.java:263]: this function is the same than the one from Driver - is there a way to have it only once and both places use it?

File: wrapper/src/main/java/software/amazon/jdbc/targetdriverdialect/TargetDriverDialectManager.java [hunk 5]
```diff
@@ -39,6 +40,10 @@ public class TargetDriverDialectManager implements TargetDriverDialectProvider {
       "wrapperTargetDriverDialect", "",
       "A unique identifier for the target driver dialect.");
 
+  public static final AwsWrapperProperty TARGET_DRIVER_AUTO_REGISTER = new AwsWrapperProperty(
+      "wrapperTargetDriverAutoRegister", "true",
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/targetdriverdialect/TargetDriverDialectManager.java:44]: let's remove the wrapper prefix

File: wrapper/src/main/java/software/amazon/jdbc/ds/AwsWrapperDataSource.java (hunk 4 above)
Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/ds/AwsWrapperDataSource.java:263]: I extracted this logic to TargetDriverHelper class.

File: wrapper/src/main/java/software/amazon/jdbc/Driver.java [hunk 6]
```diff
@@ -134,26 +135,8 @@ public Connection connect(final String url, final Properties info) throws SQLExc
...
-      if (driver == null) {
-        final List<String> registeredDrivers = Collections.list(DriverManager.getDrivers())
-            .stream()
-            .map(x -> x.getClass().getName())
-            .collect(Collectors.toList());
-        LOGGER.severe(() -> Messages.get("Driver.missingDriver", new Object[] {driverUrl, registeredDrivers}));
-        return null;
-      }
+      TargetDriverHelper helper = new TargetDriverHelper();
+      java.sql.Driver driver = helper.getTargetDriver(driverUrl, props);
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/Driver.java:139]: NIT: is `getTargetDriver` the right name for this function?
we first try to find if the driver is registered, then try to register the driver of the dialect

after this line, we either had a driver ready, either registered a new driver, either crashed
would `instantiateDriver` a better fit?

File: wrapper/src/main/java/software/amazon/jdbc/TargetDriverHelper.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.Properties;
+import java.util.stream.Collectors;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import software.amazon.jdbc.targetdriverdialect.TargetDriverDialectManager;
+import software.amazon.jdbc.util.ConnectionUrlParser;
+import software.amazon.jdbc.util.Messages;
+
+public class TargetDriverHelper {
+
+  public java.sql.Driver getTargetDriver(
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/TargetDriverHelper.java:32]: NIT: it's a really minor thing but could we add some comments in some parts of the conditional just to explain the situation and what we're doing in that case

File: wrapper/src/main/java/software/amazon/jdbc/targetdriverdialect/TargetDriverDialectManager.java [hunk 8]
```diff
@@ -121,4 +139,40 @@ private void logDialect(final String dialectCode, final TargetDriverDialect targ
         "TargetDriverDialectManager.useDialect",
         new Object[] {dialectCode, targetDriverDialect}));
   }
+
+  public boolean registerDriver(
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/targetdriverdialect/TargetDriverDialectManager.java:143]: NIT: can we add some comments on what we're doing in some conditions?

**Review by brunos-bq — APPROVED** [review]

File: wrapper/src/main/java/software/amazon/jdbc/Driver.java (hunk 6 above)
Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/Driver.java:139]: This method either return a target driver, or raises an exception. 

What could be alternatives for `getTargetDriver` method name?

**Pull request merged**